### PR TITLE
Rewrite to sidecar using CharmReconciler

### LINKED ISSUE #71: Sidecar rewrite: envoy
### Context

We rewrite all of our charms using the sidecar with base charm pattern instead of the old podspec.

### What needs to get done

Rewrite the charm using sidecar with base charm pattern.

### Definition of Done

Charm is rewritten with sidecar with base charm pattern.
All of the tests are rewritten and passing.

### PR DESCRIPTION
**Pull request by ca-scribner**
This PR refactors the PodSpec charm to sidecar using the CharmReconciler.  Changes include:
* converting all charm logic to use Components
* removing the unused grpc-web relation
* bumping charmed-kubeflow-chisme to 0.3.0

Closes #71

### CODE REVIEW

**Comment by ca-scribner** [comment]
I think the creds need to be fixed in the repo.  Will investigate tomorrow

File: tests/integration/test_charm.py
Comment (DnPlas) [inline tests/integration/test_charm.py:52]: I think we can remove it entirely, a unit test would be sufficient for testing the unit status in the absence of a required relation.

**Review by DnPlas — COMMENTED** [review]
Thanks @ca-scribner, this is looking good already, just a couple comments.

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py:51]: Yeah I'm torn on this one for the same reasons you are.  On one hand, you're right that we're never going to actually use this without the ingress.  But then we could say the same for all our dashboard charms and none of them have this requirement.

I lean toward being consistent (and not *requiring* an ingress in the charm), but I don't like that we can easily break this at bundle level

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py]: Good call.  Updated!

File: tests/integration/test_charm.py
Comment (ca-scribner) [reply tests/integration/test_charm.py:52]: Good point.  Removed.

Also, realized when I was removing this that the `build_and_deploy` never checks that envoy has actually deployed successfully.  Added that in here too

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py:76]: Sorry this was a note to myself that I didn't clean up.  I've removed the comment

What I was worried about was whether `self.grpc.component.get_data()` would get invoked when *defining* the lambda or just when calling it.  If it was invoked during definition, that might raise some odd errors that are hard to understand.  But I've now tested it to confirm `.get_data()` is only called when the lambda is called, so things work as they should.

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py:77]: You're right that `get_data()` will return empty inputs if its executed without any data on the other end.  I don't think this can happen in the charm because `self.envoy_config_generator` `depends_on=[self.grpc]`, so the grpc component must go Active before `envoy_config_generator` ever attempts to use `get_data()`.  

But, just in case it can, if `get_data()` was called and returned empty inputs, you're right that `GenerateEnvoyConfig` would break here.  In the version you reviewed, `GenerateEnvoyConfig` (and thus the charm) would go into Blocked, but not with such a helpful error message.  To improve this, I've:
* added `test_envoy_config_generator_if_grpc_get_data_returns_empty_dict` which asserts the config generator and charm are not active if malformed inputs are passed to the config generator
* changed error messaging around handling the errors so the logs explain the failure more

wdyt?

**Comment by ca-scribner** [comment]
publishing in this repo is still broken, but I’ll resolve that as a separate PR

File: src/charm.py
Comment (DnPlas) [reply src/charm.py:52]: We don't have the ingress requirement in other charms, but I don't think that is by design. [`katib-ui`](https://github.com/canonical/katib-operators/blob/main/charms/katib-ui/src/charm.py#L173) for instance is definitely using the ingress relation on every single event. I think SDI does something about missing relations (errors out?) it shows in the status message.

To me this component, alongside metadata-grpc-service (mlmd), is meant to be used by the Kubeflow Pipelines, which is also tightly coupled with other components, like the Kubeflow dashboard and the Istio ingress. That being said, and because we haven't defined a strategy for standalone components, I think we can be fine w/o requiring the ingress relation. I'd suggest to create an issue to start the discussion and log somewhere in this charm that the ingress relation is not set, if used in a Kubeflow + Istio environment, it has to be set, it can be a WARNING.

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py]: Yeah I see your point.  It felt strange introducing `metadata-envoy` here since it isn't mentioned anywhere else (at least, until we rename the charm), so I tried to reword the comment a bit instead.  wdyt?

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py]: Yeah I like discussing this as a general strategy.  There's definitely a testing benefit to letting the charm run when ingress is omitted, but then that same "feature" opens our real product up to a potential bug...  Added an issue as canonical/bundle-kubeflow#852

Comment (ca-scribner) [reply src/charm.py]: As for giving a warning if the relation is missing, I'll have to think on how to do that.

File: tests/unit/test_charm.py
Comment (ca-scribner) [reply tests/unit/test_charm.py:45]: The thought behind this was that if this component has a Blocked status then:
* the charm cannot be active (that's something the `CharmReconciler` enforces - if any `Component` is blocked, then the `CharmReconciler` puts the charm to blocked)
* the charm might be blocked by **this** component, but that is not guaranteed because some other `Component` could have blocked the charm first (for example, if we make the ingress relation required)

So I think asserting the charm is blocked by **this** component adds a flaky element to the test.  wdyt?

Thinking on this more, do we need any assertion on the `unit.status`?  If we've asserted above `harness.charm.grpc.status.message` has the error message, the only way this doesn't get to the charm is if there's a `CharmReconciler` bug.  So this assertion on `unit.status` is I think just testing the `CharmReconciler`.

File: tests/unit/test_charm.py
Comment (ca-scribner) [reply tests/unit/test_charm.py:106]: I think there's two threads to this discussion:

### Blocked vs Error

I see where you're coming from, but the `CharmReconciler` is going to prevent the charm from going into `Error` in either case.  

`CharmReconciler` catches errors thrown by any `Component` and handles them by logging and setting their `Component` to `Blocked` (so in this case, something like `BlockedStatus('[envoy_config_generator] SOMETHING')`. The motivation behind this is to prevent one erroring Component of a charm from taking down the entire charm if you don't want it to (the case in mind at the time was istio, where having an error when setting up an ingress shouldn't prevent someone from being able to enable auth on the gateway, etc)

If we want these to really throw errors to Juju, we'd have to change something in `CharmReconciler`  I think.  

### the assertion after the `config-changed`

the assertion after the unrelated `config-changed` is that the unit is **not** active.  Basically just confirming that, now that `get_data` is broken, the charm is not showing as active

File: tests/unit/test_charm.py
Comment (ca-scribner) [reply tests/unit/test_charm.py:46]: For now, I've added some assertions to be certain the `Component` is in blocked (before we just checked the message).  But we can do more changes if you like

Comment (ca-scribner) [reply tests/unit/test_charm.py:46]: Ok, added a logger to the charm and some tests.  wdyt?

File: src/charm.py
Comment (ca-scribner) [reply src/charm.py]: This warning is added now

Comment (DnPlas) [reply src/charm.py]: Sounds good.

File: tests/unit/test_charm.py
Comment (DnPlas) [reply tests/unit/test_charm.py:143]: Re blocked vs error:
In this case, though, is there anything that a user can do to unblock the charm? That's why to me it is more appropriate to error out as this is probably an issue with the charm or the related applications. If we want to avoid erroring out, because of the CharmReconciler limitations, we could put the unit in maintenance or event waiting status, but log all of the details as this error seems unrecoverable.

> the assertion after the unrelated config-changed is that the unit is not active. Basically just confirming that, now that get_data is broken, the charm is not showing as active

Got it. The comment message was a bit misleading, it says "Do something to trigger a config changed to update the charm status", I thought the unit status would've changed on a config changed event. Anyway, LGTM.

File: tests/unit/test_charm.py
Comment (DnPlas) [reply tests/unit/test_charm.py:46]: > Thinking on this more, do we need any assertion on the unit.status? If we've asserted above harness.charm.grpc.status.message has the error message, the only way this doesn't get to the charm is if there's a CharmReconciler bug. So this assertion on unit.status is I think just testing the CharmReconciler.

The assertion of the message seems right.

About how the `CharmReconciler` handles errors and unit status, I think we can have a discussion later.

File: tests/unit/test_charm.py
Comment (DnPlas) [reply tests/unit/test_charm.py:143]: Thinking about this a bit more, I see the limitations that the `CharmReconciler` has, we probably want to discuss about it, no reason to block this PR. Thanks for the clarification @ca-scribner !

File: src/charm.py
Comment (DnPlas) [reply src/charm.py]: lgtm

**Review by DnPlas — APPROVED** [review]

**Pull request merged**
